### ISSUE TITLE
On-chain fee quoting: create ISM-specific IGP contract

### ISSUE THREAD

**Issue by tkporter** [hyperlane-xyz/hyperlane-monorepo#1298]
Blocked by: https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/1296

https://github.com/hyperlane-xyz/hips/pull/3 introduces ISM-specific IGP contracts. These allow users to pass in the amount of gas their recipient's `handle` function will use, and then a stored value of the ISM-specific costs is read from the contract & added to the handle gas, and then passed to a "base" IGP as a total amount of gas. The base IGP is where the actual payment will end up and where any `GasPayment` events will be emitted.

 E.g. something like:

```solidity
contract IsmSpecificIgp is Ownable {
    IInterchainGasPaymaster baseIgp;

    // Destination domain => gas usage of the ISM.
    mapping(uint32 => uint256) ismDestinationGas;

    /**
     * @notice Accepts payment for a message, reverting if not enough payment was provided.
     * @dev The `_handleGas` should cover the gas required by the recipient's `handle` function.
     *   The gas required for ISM costs, including ISM metadata calldata and ISM.verify gas,
     *   is read from storage and is added to `_handleGas` before being passed to the baseIgp.
     * @param _messageId The message ID.
     * @param _destinationDomain The destination domain of the message.
     * @param _ismAndHandleGas The amount of gas to cover ISM metadata calldata,
     *   ISM.verify gas, and gas required by the recipient's `handle` function.
     * @param _refundAddress The address to refund any msg.value overpayment to.
     */
    function payGas(
        bytes32 _messageId,
        uint32 _destinationDomain,
        uint256 _handleGas,
        address _refundAddress
    ) external payable {
        uint256 _ismAndHandleGas = ismDestinationGas[_destinationDomain] + _handleGas;

        baseIgp.payGas{ value: msg.value }(
            _messageId,
            _destinationDomain,
            _ismAndHandleGas,
            _refundAddress
        );
    }

    // Similar logic for calculating _ismAndHandleGas and passing it into the baseIgp
    function quoteGasPayment(
      uint32 _destinationDomain,
      uint256 _gas
    ) external returns (uint256);

    function setIsmDestinationGas(uint32 _destinationDomain, uint256 _ismGas) external onlyOwner;
}
```

I think it's an open question whether we should revert if `ismDestinationGas[_destinationDomain]` is zero or not

We should:
* Write this contract, test it (adding deploy tooling for the default ISM is out of scope and done in https://github.com/hyperlane-xyz/hyperlane-monorepo/issues/1299)
